crates/rust-analyzer/src/main_loop.rs

```diff
@@ -582,53 +606,43 @@ impl GlobalState {
                     if this
                         .mem_docs
                         .insert(path.clone(), DocumentData::new(params.text_document.version))
-                        .is_some()
+                        .is_err()
                     {
                         log::error!("duplicate DidOpenTextDocument: {}", path)
                     }
-                    let changed = this
-                        .vfs
+                    this.vfs
                         .write()
                         .0
                         .set_file_contents(path, Some(params.text_document.text.into_bytes()));
-
-                    // If the VFS contents are unchanged, update diagnostics, since `handle_event`
-                    // won't see any changes. This avoids missing diagnostics when opening a file.
-                    //
-                    // If the file *was* changed, `handle_event` will already recompute and send
-                    // diagnostics. We can't do it here, since the *current* file contents might be
-                    // unset in salsa, since the VFS change hasn't been applied to the database yet.
-                    if !changed {
-                        this.maybe_update_diagnostics();
-                    }
                 }
                 Ok(())
             })?
             .on::<lsp_types::notification::DidChangeTextDocument>(|this, params| {
                 if let Ok(path) = from_proto::vfs_path(&params.text_document.uri) {
-                    let doc = match this.mem_docs.get_mut(&path) {
-                        Some(doc) => doc,
+                    match this.mem_docs.get_mut(&path) {
+                        Some(doc) => {
+                            // The version passed in DidChangeTextDocument is the version after all edits are applied
+                            // so we should apply it before the vfs is notified.
+                            doc.version = params.text_document.version;
```

Comment:
This is a little weird since technically the edits haven't been applied yet.

Reply:
Forgot to answer this!

This is interesting and subtle! Over time, I realised that it's not useful to think about program in terms of "sequence" of step. There's no "point in time". Rather, there's a discrete sequence of steps, and "current state" is meaningful only between the steps.

For example, if you have a mutex and a condvar, it doesn't matter when exacty you issue the notify call, as long as it happens within a critical section. 


Same here --the whole notification handling function is atomic with respect to changes being made.